### PR TITLE
Add hello-world and QAOA samples from Pasqal

### PR DESCRIPTION
**Pull request by ArnaudPel**
One of the final tasks required to complete the integration of Pasqal as an Azure Quantum provider was to provide sample notebooks, which is what this PR attempts to do.

### CODE REVIEW

**Comment by ArnaudPel** [comment]
@microsoft-github-policy-service agree company="Pasqal"

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (nmenetrier) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:27]:
```suggestion
    "    import pulser\n",
    "    import pulser_simulation\n",
    "except ImportError:\n",
    "    # This installs two python packages:\n",
    "    #   - pulser (core package)\n",
    "    #   - pulser_simulation\n",
    "    # If you don't need simulation in your own workflow, you may run instead:\n",
    "    #   !pip -q install pulser-core\n",
    "    # https://pulser.readthedocs.io/en/stable/installation.html\n",
    "    !pip -q install --upgrade pulser"
```
With an old pulser version on my Quantum Worskpace, I experienced an import error in the next cell.
Just added a pulser_simulation import in the try block, and the `--upgrade` parameter to the pip command.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:214]: Ah yes thanks, I had fixed that warning in the qaoa, but forgot to do the same in the hello-world.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:65]: Sure, I'll push another commit with that addition.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Great ! Thanks for having investigated and solved that.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [inline samples/hello-world/HW-pasqal-pulser.ipynb:43]: Since Pulser is a framework to construct quantum circuits (a kin to `Q#`, `Qiskit` and `Cirq`) I think we could:
1. Add another requirement to [azure-quantum](https://github.com/microsoft/azure-quantum-python/tree/main/azure-quantum), say `requirements-pulser.txt` with the following content:  
   ```
   pulser-core>=0.18,<1.0
   ```
   (assuming Pulser follows [SemVer](https://semver.org/))
The optional `requirements-pulser.txt` will get installed when user does:
   ```bash
   pip install azure-quantum[pulser]
   ```
   and it will get installed automatically when user does:
   ```bash
   pip install azure-quantum[all]
   ```
1. Remove this Notebooks cell, as the Notebooks in Azure Quantum will have all requirements installed by default.

FYI, @04diiguyi

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [inline samples/hello-world/HW-pasqal-pulser.ipynb:297]: Correct if I'm misunderstanding this, but if we don't support cost estimation for emulators and we add a cell to estimate cost for an emulator target `pasqal.sim.emu-tn` we use in this Notebook, user won't see results.
Might be we could remove this cell since the estimation won't work for this target.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:43]: Yes Pulser follows SemVer, unfortunately after digging on this subject, I find that [v < 1.0 still allows breaking changes](https://github.com/pasqal-io/Pulser/blob/develop/release.md#versioning):

> Currently (as of July 2022), Pulser is still in an early development phase and has a MAJOR of 0 - as such, both breaking and backwards compatible changes and additions to the API will be introduced through increments in the MINOR until version 1.0.0 is released.

In that light, it seems safest to specify `pulser-core=0.18`.

Other than that, happy to apply your suggestion, should I do it now, or do we need to wait for someone else's approval ?

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:297]: Indeed the output of that cell is:

>Price estimation is not implemented yet for this target.

So there is no "crash" per se, only an information message. My idea was to keep that code in the notebook so that the user learns about that azure-quantum sdk method, in the hope that they will remember to use it later, before running a job on a QPU.

As you prefer.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (a-corni) [inline samples/hello-world/HW-pasqal-pulser.ipynb:9]:
```suggestion
    "In this notebook, we'll review the basics of Azure Quantum by submitting a simple *job*, or quantum program, to [Pasqal](https://pasqal.com/). We will use [Pulser](https://pulser.readthedocs.io/) to define a quantum job that can run on Pasqal's hardware."
```

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (a-corni) [inline samples/hello-world/HW-pasqal-pulser.ipynb:24]:
```suggestion
    "Azure Quantum provides several ways to express quantum programs. In this example we are using Pulser, the quantum computing library of Pasqal to program arrays of neutral atoms. All code in this example will be written in Python.\n",
```
Taking it from https://pulser.readthedocs.io/en/stable/ ;)

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (a-corni) [inline samples/hello-world/HW-pasqal-pulser.ipynb:150]:
```suggestion
    "Let's create a simple Pulser Sequence to run on an array of neutral atoms (a Register)."
```

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:487]:
```suggestion
    "However, using QAOA to solve the problem is not the best idea; it's difficult to yield a >90% quality solution without going to high depths of the QAOA, implying that the growing closed-loop optimization can rapidly become expensive, with no guarantee of convergence. 
    We therefore propose to investigate another approach called the Quantum Adiabatic Algorithm (QAA). This method is exclusively analog and can be well implemented on Pasqal's Fresnel machines that are tailored to solve analog quantum jobs. It is fast, reliable and shows optimal convergence to the solution."
```

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:509]: Do you really want to use DigitalAnalogDevice ? I think AnalogDevice would suit better what you are trying to prove.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:589]: nit: 
```suggestion
    "for T in np.linspace(1000, 10000, 10):\n",
```
Should work as well, and is perhaps cleaner ? I let you see !

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:590]: The demonstration would be better with AnalogDevice, but the value of T might have to be restrained to `np.linspace(1000, 5000, 10)` (I think the max duration of the Rydberg channel in AnalogDevice is 5000ns)

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:820]: So we don't show how to run it on the QPU ? If you replace DigitalAnalogDevice by AnalogDevice and the values of T by the values I provided above, you should have something that can run on Fresnel. I let you see if that's something of interest.

**Review by a-corni — COMMENTED** [review]
Looking good ! I am just letting a couple of comments on syntax and QAOA :)

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:510]: The sample I received from Anne-Claire was using Chadoq2, which was triggering the following warning in the console:

>DeprecationWarning: The 'Chadoq2' device has been deprecated. For a similar device combining global and local addressing, consider using `DigitalAnalogDevice`.

So I applied that suggestion :grinning:

Since I'm not the author of the notebook (just integrating it here), I suppose you are more knowledgable than me @a-corni.

Your call.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:590]: As mentioned above, since I'm not the author, I would feel more comfortable leaving the code as is, unless you are sure :wink:

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:591]: So I'm not confident about taking the risk of breaking the notebook, is it ok to leave as it is ?

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:821]: I could be a good idea to write another cell with markdown or commented code to explain how to adapt this notebook for Pasqal QPU usage yes, as you mention above.

Given the price of a QPU hour though, I would not recommend adding "runnable QPU code" in these sample notebooks, to make sure the user doesn't get a bad surprise while trying to learn.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [reply samples/hello-world/HW-pasqal-pulser.ipynb:43]: @ArnaudPel the `pulser-core==0.18` sounds good. We could also use `pulser-core==0.18.*` if we're certain no breaking change will be introduced with a patch.
Could you please make the aforementioned changes? Please, let us know if any help needed.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:487]: Could you please split the string so it would be a valid JSON-array item?

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:101]: Yes I initially wanted to use the same code as other HW notebooks, then I realized that `AzureQuantumService` is imported from `cirq`, which does not seem usable in the Pasqal/pulser environment.

Trying to dive into the code, I see that the implementation of `service.targets` relies on a `TargetFactory()`, should I implement another factory for Pasqal ?

I would welcome some guidance on this :bow:

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: I have added the file `requirements-pulser.txt` in this PR, and consequently removed this cell.

I did not try to test that locally, is there a way to do so before this PR is merged ?

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: I have applied your latest suggestion, showing cost estimates for fresnel.

Sadly, I receive the same message (not implemented) :pensive:

I think I need to open an issue on our side to implement at least that.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Line removed.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Oh right, link updated, thanks !

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Oh, I think the issue comes from the suggestion I applied directly from Gitlab.

Should be fixed now.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Sure, you could:
1. Create a new Python environment (Conda, venv), 
2. Navigate to `azure-quantum` folder
3. Run 
```python
pip install -e .[pulser]
```
4. Run code from the notebook you put together

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Oh, if PASQAL doesn't yet have cost estimation implemented for any of the targets, might be we could remove this cell altogether for now. Like I mentioned above, we don't typically advertise cells that are work-in-progress

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (kikomiss) [reply samples/hello-world/HW-pasqal-pulser.ipynb:87]: Sorry for the confusion, I shared a wrong example. We don't have to implement a custom Target to list available targets. This code will work:
```python
print("This workspace's targets:")
for target in workspace.get_targets():
    print("-", target.name)
```

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:512]: It's okay, let's go with this solution :)

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:592]: I am pretty sure it works but I am okay with the current state, just being picky no worries

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:593]: It's okay to leave it as it is

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (a-corni) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:823]: Okay it was just a suggestion no worries ;)

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Ah yes, after a bit of head scratching I figured out that I needed to wrap `.[pulser]` around quotes, then it worked like a charm :relieved: 

So I confirm that this approach seems to work locally.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Sure, I removed step 5 completely, and opened an issue on our side.

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb:87]: Great ! That seems to work as expected :raised_hands:

Should now be fixed.

**Comment by ArnaudPel** [comment]
FYI while re-testing the HW notebook this morning, I faced a new error which I was not seeing ~15h ago:

>'DIGITALANALOGDEVICE' is not a valid DeviceTypeName

We are investigating this internally.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:29]: Could we now remove this cell since we implemented `pip install azure-quantum[pulser]` and the Pulser will be installed automatically in Notebooks?

Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Yes sorry I briefly thought about it then forgot. Cell removed.

**Comment by ArnaudPel** [comment]
> FYI while re-testing the HW notebook this morning, I faced a new error which I was not seeing ~15h ago:
> 
> > 'DIGITALANALOGDEVICE' is not a valid DeviceTypeName
> 
> We are investigating this internally.

This was due to an incorrect target being used, following the application of one of the suggestions. Now fixed by reinstating

    target = workspace.get_targets(name="pasqal.sim.emu-tn")

Thanks @nmenetrier !

**Comment by ArnaudPel** [comment]
@kikomiss I think that all threads have been addressed, we might be good to go :grinning: :crossed_fingers:

**Comment by ArnaudPel** [comment]
Hello ! Sending a little reminder to see if we can merge this PR :grinning:

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:683]: `target.latest_session` always gets set to the newly opened session under the hood. Was there a particular reason we needed explicitly set `target.latest_session` here?

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:682]: nit: could you please use 
```python
with target.open_session(name="QUBO Problem", job_failure_policy=SessionJobFailurePolicy.CONTINUE) as session
```
instead of explicit `session.close()`. In this case we can combine this cell with the one below and move the cell "Note: the next cell may take up to 5 minutes to complete." above this one.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:681]: super-nit: import of `Session` is not used and could be removed.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (kikomiss) [inline samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:735]: Is there an opportunity for us to add a summery at the end and explain the difference of results produced by `QutipEmulator` and PASQAL emulator?
For instance:
- Why are they different
- Which emulator has higher fidelity and why
- Which emulator should be used in which case (trade-offs like `QutipEmulator` is quicker to produce the result, but is less accurate, `QutipEmulator` doesn't converge is certain cases therefore couldn't be used, etc)

**Comment by kikomiss** [comment]
> Hello ! Sending a little reminder to see if we can merge this PR 😀

@ArnaudPel I apologize, for the delay with review. I've left a couple of `nit` comments and a suggestion on [qaoa-for-quadratic-unconstrained-binary-optimization.ipynb](https://github.com/microsoft/azure-quantum-python/pull/607/files/05451e1bdcfa8e096ec531daceab7d125b514013#diff-5bc21485c6191adb934748bf4ee5eb88bb1367f616c9d1b434cbb9a69d9c3c07).

File: samples/hello-world/HW-pasqal-pulser.ipynb
Comment (ArnaudPel) [reply samples/hello-world/HW-pasqal-pulser.ipynb]: Ah yes sorry I missed that typo in the comment.

I have applied both your suggestions.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Indeed, sorry for not having spotted that, I have updated `requirements-pulser.txt`.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Done.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: As yes I don't know why it was 10 here, I have increased to 100.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: I have used the context manager and split the cells a bit differently, I hope you like it.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb]: Well spotted, I have removed it.

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:728]: I would need to loop back internally to answer these questions, and I'm a bit worried about how long that could take.

Would it be possible to merge this PR, and refine this point in a subsequent PR ?

Comment (kikomiss) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:728]: Sounds good! Let's merge it and do a follow up PR later. Sorry for so much churn on this PR and thank you for working with me on improving it!

**Review by kikomiss — APPROVED** [review]

File: samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb
Comment (ArnaudPel) [reply samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb:728]: Thank you very much :relieved:

I have requested this to be prioritized on our side.

**Pull request merged**
